src/aspire/utils/rotation.py

```diff
@@ -0,0 +1,289 @@
+"""
+Define a Rotation Class for customized rotation operations used by ASPIRE.
+"""
+
+import copy
+import numpy as np
+from numpy.linalg import norm
+from scipy.linalg import svd
+from scipy.spatial.transform import Rotation as sp_rot
+
+from aspire.utils import ensure
+from aspire.utils.matlab_compat import Random
+
+
+class Rotation:
+    """
+    Define a Customized Rotation class
+    """
+
+    def __init__(self, num_rots, seed=None, seq='ZYZ',
+                 angles=None, degrees=False, dtype=np.float32):
+        """
+        Initialize a Rotation object
+
+        :param num_rots: Total number of rotation sets
+        :param seed: Seed for initializing random rotations
+        :param seq:  Sequence of order applying Euler angles
+        :param angles: Euler angles to generate rotation matrices
+        :param degrees: Whether Euler angles in degrees or not
+        :param dtype:   data type for angles and rotation matrices
+        """
+        self.rot_seq = seq
+        self.degrees = degrees
+        self.dtype = np.dtype(dtype)
+        if angles:
+            ensure(num_rots == angles.shape[0],
+                   'Number of rotation matrices should equal to '
+                   'number of sets of Euler angles.')
+            self.angles = angles
+        else:
+            self.angles = self._uniform_random_angles(
+                num_rots, seed=seed).astype(self.dtype)
+        self.num_rots = num_rots
+
+        self.data = self.rot_matrices.astype(self.dtype)
+        self.shape = (num_rots, 3, 3)
+
+    def _uniform_random_angles(self, n, seed=None):
+        """
+        Generate random 3D rotation angles
+
+        :param n: The number of rotation angles to generate
+        :param seed: Random integer seed to use. If None,
+            the current random state is used.
+        :return: A n-by-3 ndarray of rotation angles
+        """
+        # Generate random rotation angles, in radians
+        angles = np.zeros((n, 3), dtype=self.dtype)
+        with Random(seed):
+            angles = np.column_stack((
+                np.random.random(n) * 2 * np.pi,
+                np.arccos(2 * np.random.random(n) - 1),
+                np.random.random(n) * 2 * np.pi
+            ))
+        return angles
+
+    @property
+    def rot_matrices(self):
+        """
+        :return: Rotation matrices as a n x 3 x 3 array
+        """
+        return self._rotations.as_matrix()
+
+    @rot_matrices.setter
+    def rot_matrices(self, values):
+        """
+        Set rotation matrices
+
+        :param values: Rotation matrices as a n x 3 x 3 array
+        :return: None
+        """
+        self._rotations = sp_rot.from_matrix(values)
+        self.data = values.astyp(self.dtype)
+
+    @property
+    def angles(self):
+        """
+        :return: Rotation angles in radians, as a n x 3 array
+        """
+        return self._rotations.as_euler(self.rot_seq)
+
+    @angles.setter
+    def angles(self, values):
```

Comment:
If `self.degree`, assume input is in degrees.

Reply:
Changed to use degrees in default.